File: lib/models/food.dart
// class Food {
//   final String id;
//   final String name;
//   final String category;
//   final String description;
//   final double price;
//   final double? originalPrice;
//   final double? discountPercent;
//   final String image;
//   final bool isPopular;
//   final bool isSpecialOffer;
//   final double rating;
//   final int totalReviews;
//   final List<String> ingredients;
//   final bool available;

//   Food({
//     required this.id,
//     required this.name,
//     required this.category,
//     required this.description,
//     required this.price,
//     this.originalPrice,
//     this.discountPercent,
//     required this.image,
//     this.isPopular = false,
//     this.isSpecialOffer = false,
//     this.rating = 0.0,
//     this.totalReviews = 0,
//     this.ingredients = const [],
//     this.available = true,
//   });

//   factory Food.fromJson(Map<String, dynamic> json) {
//     return Food(
//       id: json['_id'] ?? json['id'] ?? '',
//       name: json['name'] ?? '',
//       category: json['category'] ?? '',
//       description: json['description'] ?? '',
//       price: (json['price'] ?? 0.0).toDouble(),
//       originalPrice: json['originalPrice']?.toDouble(),
//       discountPercent: json['discountPercent']?.toDouble(),
//       image: json['image'] ?? '',
//       isPopular: json['isPopular'] ?? false,
//       isSpecialOffer: json['isSpecialOffer'] ?? false,
//       rating: (json['rating'] ?? 0.0).toDouble(),
//       totalReviews: json['totalReviews'] ?? 0,
//       ingredients: List<String>.from(json['ingredients'] ?? []),
//       available: json['available'] ?? true,
//     );
//   }

//   bool get isLocalImage {
//     return image.startsWith('assets/');
//   }

//   String get displayImage {
//     return image;
//   }
// }
class Food {
  final String id;
  final String name;
  final String category;
  final String description;
  final double price;
  final double? originalPrice;
  final double? discountPercent;
  final String image;
  final bool isPopular;
  final bool isSpecialOffer;
  final double rating;
  final int totalReviews;
  final List<String> ingredients;
  final bool available;

  Food({
    required this.id,
    required this.name,
    required this.category,
    required this.description,
    required this.price,
    this.originalPrice,
    this.discountPercent,
    required this.image,
    this.isPopular = false,
    this.isSpecialOffer = false,
    this.rating = 0.0,
    this.totalReviews = 0,
    this.ingredients = const [],
    this.available = true,
  });

  factory Food.fromJson(Map<String, dynamic> json) {
    // Handle both MongoDB '_id' and Firebase 'id'
    String foodId;
    if (json['_id'] != null) {
      foodId = json['_id'].toString();
    } else if (json['id'] != null) {
      foodId = json['id'].toString();
    } else {
      foodId = '';
    }

    return Food(
      id: foodId,
      name: json['name'] ?? '',
      category: json['category'] ?? '',
      description: json['description'] ?? '',
      price: (json['price'] ?? 0.0).toDouble(),
      originalPrice: json['originalPrice']?.toDouble(),
      discountPercent: json['discountPercent']?.toDouble(),
      image: json['image'] ?? '',
      isPopular: json['isPopular'] ?? false,
      isSpecialOffer: json['isSpecialOffer'] ?? false,
      rating: (json['rating'] ?? 0.0).toDouble(),
      totalReviews: json['totalReviews'] ?? 0,
      ingredients: List<String>.from(json['ingredients'] ?? []),
      available: json['available'] ?? true,
    );
  }

  bool get isLocalImage {
    return image.startsWith('assets/');
  }

  String get displayImage {
    return image;
  }

  // Calculate discounted price
  double get discountedPrice {
    if (isSpecialOffer && originalPrice != null && discountPercent != null) {
      return originalPrice! - (originalPrice! * discountPercent! / 100);
    }
    return price;
  }

  // Check if item is on discount
  bool get isOnDiscount {
    return isSpecialOffer && originalPrice != null && discountPercent != null;
  }

  // Calculate savings
  double? get youSave {
    if (isOnDiscount) {
      return originalPrice! - discountedPrice;
    }
    return null;
  }
}

File: server/models/Food.js
const mongoose = require('mongoose');

const reviewSchema = new mongoose.Schema({
    user: {
        type: String,
        required: true
    },
    rating: {
        type: Number,
        required: true,
        min: 1,
        max: 5
    },
    comment: {
        type: String,
        required: true
    },
    // date: {
    //     type: Date,
    //     default: Date.now
    // }
});

const foodSchema = new mongoose.Schema({
    name: {
        type: String,
        required: true,
        trim: true
    },
    category: {
        type: String,
        required: true,
        enum: ['burger', 'pizza', 'cake', 'icecream', 'softdrink', 'burrito']
    },
    description: {
        type: String,
        required: true
    },
    price: {
        type: Number,
        required: true,
        min: 0
    },
    originalPrice: {
        type: Number,
        min: 0
    },
    discountPercent: {
        type: Number,
        min: 0,
        max: 100
    },
    image: {
        type: String,
        required: true
    },
    isPopular: {
        type: Boolean,
        default: false
    },
    isSpecialOffer: {
        type: Boolean,
        default: false
    },
    rating: {
        type: Number,
        default: 0,
        min: 0,
        max: 5
    },
    totalReviews: {
        type: Number,
        default: 0
    },
    ingredients: {
        type: [String],
        default: []
    },
    reviews: [reviewSchema],
    available: {
        type: Boolean,
        default: true
    }
}, {
    timestamps: true
});

// Virtual for discounted price
foodSchema.virtual('discountedPrice').get(function() {
    if (this.originalPrice && this.discountPercent) {
        return this.originalPrice - (this.originalPrice * this.discountPercent / 100);
    }
    return this.price;
});

// Method to calculate average rating
foodSchema.methods.calculateAverageRating = function() {
    if (this.reviews.length === 0) return 0;
    const sum = this.reviews.reduce((acc, review) => acc + review.rating, 0);
    return sum / this.reviews.length;
};

const Food = mongoose.model('Food', foodSchema);
module.exports = Food;